### PR TITLE
Dev weighted encoders

### PR DESCRIPTION
**Pull request by apyskir**
Weighted loss across pixels implemented and different encoders merged
* preparation of files containing distances to the closest objects from each pixels
* dedicated loader for reading distance files
* reflection padding after augmentations added
* NeptuneMonitor adapted for handling modified targets
* weighted loss implemented: pixels close to target objects are given higher weight
* encoders adapted by @taraspiotr included
* label() was moved from postprocessing.py to utils.py

### CODE REVIEW

File: callbacks.py [hunk 1]
```diff
@@ -56,6 +56,9 @@ def get_prediction_masks(self):
             X = data[0]
             targets_tensors = data[1:]
 
+            if (targets_tensors[0].size()[1] > 1):
+                targets_tensors = [target_tensor[:, :1, :, :] for target_tensor in targets_tensors]
```
Comment (jakubczakon) [inline callbacks.py]: @apyskir I am pretty sure that python3 lets you do `target_tensor[: , :1]` here.

File: loaders.py [hunk 2]
```diff
@@ -9,6 +9,8 @@
 from attrdict import AttrDict
 from skimage.transform import resize
 from torch.utils.data import Dataset, DataLoader
+from sklearn.externals import joblib
+from cv2 import resize as resize_cv
```
Comment (jakubczakon) [inline loaders.py]: @apyskir why exactly do we want 2 resize flavours?

File: loaders.py [hunk 3]
```diff
@@ -68,6 +70,70 @@ def __getitem__(self, index):
...
+        return self.X.shape[0]
+
+    def __getitem__(self, index):
+        img_filepath = self.X[index]
+        Xi = self.load_image(img_filepath)
+
+        if self.y is not None:
+            mask_filepath = self.y[index]
+            Mi = self.load_image(mask_filepath)
+            distance_filepath = mask_filepath.replace("/masks/", "/distances/")[:-4]
```
Comment (jakubczakon) [inline loaders.py]: @apyskir I presume this action is performed to drop `.jpg` .In the future we may have a competition with `.jpeg` images or something. I think you can call `from os.path import basename; basename(filename)`

File: loaders.py [hunk 4]
```diff
@@ -68,6 +70,70 @@ def __getitem__(self, index):
...
+            Di = self.load_distances(distance_filepath)
+            Di = np.sum(Di, axis=2).astype(np.uint8)  # TODO: remove it when Di will be sum of distances to 2 closest objects
```
Comment (jakubczakon) [inline loaders.py]: @apyskir Generally I'd prefer no TODO in PR but in this case I understand. Drop an issue with this small refactor not to forget later.

File: loaders.py [hunk 5]
```diff
@@ -68,6 +70,70 @@ def __getitem__(self, index):
...
+
+            if self.train_mode and self.image_augment_with_target is not None:
+                Xi, Mi = from_pil(Xi, Mi)
+                Xi, Mi, Di = self.image_augment_with_target(Xi, Mi, Di)
+                if self.image_augment is not None:
+                    Xi = self.image_augment(Xi)
+                Xi, Mi, Di = to_pil(Xi, Mi, Di)
+
+            if not self.train_mode:
+                Di = to_pil(Di)
```
Comment (jakubczakon) [inline loaders.py]: @apyskir why do we need Di in the inference mode?

File: callbacks.py (hunk 1 above)
Comment (apyskir) [reply callbacks.py]: @jakubczakon Yes, you're right.

File: loaders.py (hunk 3 above)
Comment (apyskir) [reply loaders.py]: @jakubczakon `os.path.splitext(filename)[0]` is what actually does the job

File: loaders.py (hunk 4 above)
Comment (apyskir) [reply loaders.py]: @jakubczakon done

File: loaders.py (hunk 5 above)
Comment (apyskir) [reply loaders.py]: @jakubczakon actually we don't. But look here https://github.com/minerva-ml/open-solution-mapping-challenge/blob/dev/loaders.py#L97 This suggests, that we need mask in inference mode, but we don't.
Am I right, that when `self.y!=None` then `self.train_mode==True`? Maybe these parameters are equivalent, but sometimes first is used, sometimes the other.

File: models.py [hunk 6]
```diff
@@ -146,3 +251,36 @@ def callbacks_unet(callbacks_config):
...
                    ])
+
+
+def multiclass_weighted_segmentation_loss(output, target, w0, sigma):
+    '''
+    w1 is temporarily torch.ones - it should handle class imbalance for thw hole dataset
+    '''
+    distance = target[:, 1, :, :]
+    target = target[:, 0, :, :].long()
+    w1 = Variable(torch.ones(distance.size()), requires_grad=False)  # TODO: fix it to handle class imbalance
```
Comment (jakubczakon) [inline models.py]: @apyskir add this as issue so we don't forget

File: models.py [hunk 7]
```diff
@@ -146,3 +251,36 @@ def callbacks_unet(callbacks_config):
...
+    weights = get_weights(distance, w1, w0, sigma)
+    torch_softmax = torch.nn.Softmax2d()
+    probabilities = torch_softmax(output)
+    negative_log_likelihood = torch.nn.NLLLoss2d(reduce=False)
+    loss_per_pixel = negative_log_likelihood(torch.log(probabilities), target)
+    loss = torch.mean(loss_per_pixel * weights)
+    return loss
+
+
+def get_weights(d, w1, w0, sigma):
```
Comment (jakubczakon) [inline models.py]: @apyskir this is private function so I would go with _get_weights I think

File: models.py [hunk 8]
```diff
@@ -146,3 +251,36 @@ def callbacks_unet(callbacks_config):
...
+    weights = w1 + w0 * torch.exp(-(d ** 2) / (sigma ** 2))
+    weights[d == 0] = 1
+    return weights
+
+
+def get_loss_params(w0, sigma):
```
Comment (jakubczakon) [inline models.py]: @apyskir this is also private -> _get_loss_params

File: pipelines.py [hunk 9]
```diff
@@ -26,29 +26,28 @@ def unet(config, train_mode):
...
+                  adapter={'y_pred': ([(mask_postprocessed.name, 'images')]),
+                           'y_scores': ([(mask_postprocessed.name, 'scores')])
                            },
                   cache_dirpath=config.env.cache_dirpath,
                   save_output=save_output,
                   load_saved_output=False)
     return output
 
 
+def unet_weighted(config, train_mode):
```
Comment (jakubczakon) [inline pipelines.py]: @apyskir I really like that you used the .get_step method here. Shows how reusable steps can be.

File: loaders.py (hunk 2 above)
Comment (apyskir) [reply loaders.py]: `skimage.transform.resize` accepts only numbers [-1,1]. Distances don't match this requirement. Maybe I could divide the distance matrix `Di` by `Di.max()` and after resize multiply by this number, but I am not sure if it is OK. I prefer as little resize/rescale operations on `Di` as possible

File: loaders.py (hunk 5 above)
Comment (jakubczakon) [reply loaders.py]: @apyskir I see train_mode with validation data still needs to perform transformations on target (not augmentations). That is why this weird situation arises. Nvmd just clumsy coding on my part. I will add an issue

File: models.py (hunk 7 above)
Comment (apyskir) [reply models.py]: @jakubczakon OK

File: pipelines.py (hunk 9 above)
Comment (apyskir) [reply pipelines.py]: @jakubczakon Yes, I was so happy when I figured out how easily I can change 2 steps in `unet` pipeline. This case (or similiar) could be part of a notebook showing how to use steps.

File: models.py (hunk 6 above)
Comment (apyskir) [reply models.py]: @jakubczakon Done
